interfaces/devdoc/file_requirements.md

```diff
@@ -0,0 +1,154 @@
+
+**SRS_FILE_43_040: [** If `position` + `size` is greater than `INT64_MAX`, then `file_write_async` shall fail and return `FILE_WRITE_ASYNC_INVALID_ARGS`. **]**
+
+**SRS_FILE_43_012: [** If `user_callback` is `NULL` then `file_write_async` shall fail and return `FILE_WRITE_ASYNC_INVALID_ARGS`. **]**
+
+**SRS_FILE_43_014: [** `file_write_async` shall enqueue a write request to write `source`'s content to the `position` offset in the file. **]**
+
+**SRS_FILE_43_041: [** If `position` + `size` is greater than the size of the file and the call to write is succesfull, `file_write_async` shall grow the file to accomodate the write. **]**
+
+**SRS_FILE_43_035: [** If the call to write the file fails, `file_write_async` shall fail and return `FILE_WRITE_ASYNC_WRITE_ERROR`. **]**
+
+**SRS_FILE_43_008: [** `file_write_async` shall call `user_call_back` passing `user_context` and `success` depending on the success of the asynchronous write operation.**]**
+
+**SRS_FILE_43_015: [** If there are any failures, `file_write_async` shall fail and return `FILE_WRITE_ASYNC_ERROR`. **]**
+
+**SRS_FILE_43_030: [** `file_write_async` shall succeed and return `FILE_WRITE_ASYNC_OK`. **]**
+
+## file_read_async
+
+```c
+MOCKABLE_FUNCTION_WITH_RETURNS(, FILE_READ_ASYNC_RESULT, file_read_async, FILE_HANDLE, handle, uint32_t, size, uint64_t, position, FILE_READ_CB, user_callback, void*, user_context)(FILE_READ_ASYNC_OK, FILE_READ_ASYNC_ERROR);
+```
+
+`file_read_async` issues an asynchronous read request.
+
+**SRS_FILE_43_017: [** If `handle` is `NULL` then `file_read_async` shall fail and return `FILE_READ_ASYNC_INVALID_ARGS`. **]**
+
+**SRS_FILE_43_020: [** If `user_callback` is `NULL` then `file_read_async` shall fail and return `FILE_READ_ASYNC_INVALID_ARGS`. **]**
+
+**SRS_FILE_43_032: [** `file_read_async` shall allocate `size` bytes named `destination` for reading the data. **]**
+
+**SRS_FILE_43_021: [** `file_read_async` shall enqueue a read request to read `handle`'s content at `position` offset and write it to `destination`. **]**
+
+**SRS_FILE_43_036: [** If the call to read the file fails, `file_read_async` shall fail and return `FILE_READ_ASYNC_READ_ERROR`. **]**
+
+**SRS_FILE_43_039: [** If `position` + `size` exceeds the size of the file, `user_callback` shall be called with `success` as `false`. **]**
+
+**SRS_FILE_43_016: [** `file_read_async` shall call `user_callback` passing `user_context` and `success` depending on the success of the asynchronous read operation.**]**
+
+**SRS_FILE_43_022: [** If there are any failures then `file_read_async` shall fail and return `FILE_READ_ASYNC_ERROR`. **]**
+
+**SRS_FILE_43_031: [** `file_read_async` shall succeed and return `FILE_READ_ASYNC_OK`. **]**
+
+## file_extend
+
+```c
+MOCKABLE_FUNCTION_WITH_RETURNS(, int, file_extend, FILE_HANDLE, handle, uint64_t, desired_size)(0, MU_FAILURE);

```

Comment:
>file_extend [](start = 38, length = 11)

To be useable, the module needs to provide a get_file_size.

Here's why:

The module is intended to be used by BSDL. To create a log of "X" size at the moment only the following 2 APIs can be called:
line1: file_create(xxxx)
line2: file_extend(LOG_SIZE_FROM_SERVICE_FABRIC_CONFIG)

At line 2, there are 3 possibilities
1) the file now has 0 bytes (brand new created) => go ahead and extend it, BSDL is happy.
2) file file now has between [1... LOG_SIZE_FROM_SERVICE_FABRIC_CONFIG-1] => do nothing, don't open the replica, likely a previous incarnation of BSDL was created there, and making the file grow larger would lose data (we don't know how to extend BSDL files because we don't know how to account for the new extra space - our log is circular). BSDL unhappy
3) file has exactly LOG_SIZE_FROM_SERVICE_FABRIC_CONFIG => GREAT! proceed to open! BSDL happy!
4) File has more than LOG_SIZE_FROM_SERVICE_FABRIC_CONFIG => do nothing, BSDL should not open, truncating the file means possibly truncating information => data loss.

So a "get current file size" API needs to exist in the module.

Reply:
BSDL doesn't do any of that today right? This is to address that task we have that basically says BSDL shall not change the size of the log if it has valid data?

---
In reply to: [452977250](https://github.com/Azure/azure-c-pal/pull/7#discussion_r452977250) [](ancestors = 452977250)

Reply:
What BSDL does today... errr.... I think it will refuse to extend / shrink an existing file by some very convoluted mechanism. I think that mechanism might not be entirely contained in BSDL either. (see bs_disk_placement... in case of SF. No replication might end up extending... ). However, that is slightly irrelevant for what we're trying to achieve here - that is - provide a PAL file API. Oh well.

I enumerated the above 4 cases to explain why we need a getfilesize - without it... data integrity can be lost.

I would entirely remove at this moment all mentions of file_extend from here. set/get_file_size can come in a completely separated PR.

---
In reply to: [452980688](https://github.com/Azure/azure-c-pal/pull/7#discussion_r452980688) [](ancestors = 452980688,452977250)